### DISCUSSION TITLE
Support for negative conditional tokens

### DISCUSSION THREAD

**Discussion by bschrag620** [yippee-fun/phlex#385] [Ideas]
I really like the `tokens` and `classes` helpers. One thing I ran into is there are cases where under the true condition I need a token added, and under a false condition I need a different token added. Currently that might look something like this:
```
class View < ApplicationView
  def template
    div(class: tokens(true_case?: 'foo', false_case?: 'bar'))
  end

  def true_case?
  end

  def false_case?
    !true_case?
  end
end
```

The idea would be to utilize an array to hold the opposite value. I realize arrays are already a supported conditional token value so it would mean not supporting an array as a value for conditional tokens or allowing nested arrays to support the negative case:

```
class View < ApplicationView
  def template
    # idea here is when true_case? is false ['bar'] is selected
    div(class: tokens(true_case?: [['foo'], ['bar']]))
  end

  def true_case?
  end
end
```

I think there is an argument to be made about supporting a hash token where the return value from the conditional check is used as a lookup in the hash:
```
    div(class: tokens(value_method: {value_1: 'foo', value_2: 'bar', value_3: 'foobar'})
```

The refactor that lives in my head would involve creating a `Phlex::TokenGenerator` class, which `token` would instantiate.

**Comment by joeldrapper** [comment]
I really like this idea. Perhaps we could use a `Hash` instead? I’m thinking something like this.

```ruby
div(
  class: tokens(
    active?: { then: "active", else: "inactive" }
  )
)
```

The `Array` is more concise and it matches the order of a ternary operator, but I think a `Hash` with `then` / `else` keys would pretty clear to read for people who are new to the framework, while the `Array` would definitely need to be learned.

The idea of using the method’s return value as a `Hash` look-up key is interesting, but I don't think it’s necessary as it's quite easy to implement yourself.

```ruby
COLORS = {...}
SIZES = {...}
STYLES = {...}

div(
  class: tokens(COLORS[@color], STYLES[@style], SIZES[@size])
)
```

Or inline

```ruby
div(
  class: tokens(
    {
      a: "foo",
      b: "bar"
    }[style]
  )
)

def style = :a
```

**Reply by bschrag620** [reply]
I can see the benefit to `then`/`else` as far as readability and ease of learning. Also a very valid point regarding implementing the hash lookup. I can play with this a bit this weekend and come up with something.

**Reply by bschrag620** [reply]
Submitted: https://github.com/joeldrapper/phlex/pull/392
